packages/react-form-renderer/src/form-renderer/form-renderer.js

```diff
@@ -1,77 +1,132 @@
-import React, { useState, useRef } from 'react';
-import Form from '../form';
 import arrayMutators from 'final-form-arrays';
-import PropTypes from 'prop-types';
 import createFocusDecorator from 'final-form-focus';
+import PropTypes from 'prop-types';
+import React, {useCallback, useMemo, useRef, useState} from 'react';
 
+import defaultSchemaValidator from '../default-schema-validator';
+import defaultValidatorMapper from '../validator-mapper';
+import Form from '../form';
 import RendererContext from '../renderer-context';
 import renderForm from './render-form';
-import defaultSchemaValidator from '../default-schema-validator';
 import SchemaErrorComponent from './schema-error-component';
-import defaultValidatorMapper from '../validator-mapper';
+
+
+const isFunc = (fn) => (typeof fn === 'function');
 
 const FormRenderer = ({
+  actionMapper,
+  children,
+  clearedValue,
+  clearOnUnmount,
   componentMapper,
+  decorators,
   FormTemplate,
-  onSubmit,
+  FormTemplateProps,
+  mutators,
   onCancel,
+  onError,
   onReset,
-  clearOnUnmount,
-  subscription,
-  clearedValue,
+  onSubmit,
   schema,
-  validatorMapper,
-  actionMapper,
   schemaValidatorMapper,
+  subscription,
+  validatorMapper,
   ...props
 }) => {
   const [fileInputs, setFileInputs] = useState([]);
   const registeredFields = useRef({});
   const focusDecorator = useRef(createFocusDecorator());
-  let schemaError;
+  const validatorMapperMerged = useMemo(() => ({
+    ...defaultValidatorMapper,
+    ...validatorMapper
+  }), [validatorMapper]);
+  const mutatorsMerged = useMemo(() => ({
+    ...arrayMutators,
+    ...mutators
+  }), [mutators]);
+  const decoratorsMerged = useMemo(() => ([
+    focusDecorator.current,
+    ...(Array.isArray(decorators) ? decorators : [])
+  ]), [decorators]);
+
+  const handleSubmitCallback = useCallback((values, formApi, ...args) => {
+    return !isFunc(onSubmit) ? undefined : onSubmit(values, {...formApi, fileInputs}, ...args);
+  }, [onSubmit, fileInputs]);
+
+  const handleCancelCallback = useCallback((getState) => (...args) => {
+    return !isFunc(onCancel) ? void 0 : onCancel(getState().values, ...args);
+  }, [onCancel]);
+
+  const handleResetCallback = useCallback((reset) => (...args) => {
+    reset();
+    return !isFunc(onReset) ? void 0 : onReset(...args);
+  }, [onReset]);
 
-  const setRegisteredFields = (fn) => (registeredFields.current = fn({ ...registeredFields.current }));
-  const internalRegisterField = (name) => {
-    setRegisteredFields((prev) => (prev[name] ? { ...prev, [name]: prev[name] + 1 } : { ...prev, [name]: 1 }));
-  };
+  const handleErrorCallback = useCallback((...args) => {
+    // eslint-disable-next-line no-console
+    console.error(...args);
```

Comment:
The original code is calling the `console.log` with just the error message without the whole stack
```js
  try {
    const validatorTypes = Object.keys(validatorMapperMerged);
    const actionTypes = actionMapper ? Object.keys(actionMapper) : [];
    defaultSchemaValidator(schema, componentMapper, validatorTypes, actionTypes, schemaValidatorMapper);
  } catch (error) {
    schemaError = error;
    // eslint-disable-next-line no-console
    console.error(error);
    // eslint-disable-next-line no-console
    console.log('error: ', error.message);
  }
  ```
That is why the second test is falling. I am honestly OK with just using the `console.error`. @rvsia do you have any objections to removing the `console.log` part? If no we can just drop the expect from the test

Reply:
@rvsia ?

Reply:
@zgover sorry I was on vacation. Just delete the console log check from tests. It's not necessary to log the same message multiple times.